python/ray/autoscaler/_private/kuberay/node_provider.py

```diff
@@ -0,0 +1,234 @@
+        return make_node_tags(data["metadata"]["labels"], status_tag(data))
+
+    def non_terminated_nodes(self, tag_filters: Dict[str, str]) -> List[str]:
+        """Return a list of node ids filtered by the specified tags dict.
+        Also updates caches of ips and tags.
+        """
+        data = self._get("/api/v1/namespaces/default/pods/")
+
+        logger.info("Called non_terminated_nodes with tag_filters {}".format(tag_filters))
+
+        result = []
+        for pod in data["items"]:
+            labels = pod["metadata"]["labels"]
+            if labels.get("ray.io/cluster") == "raycluster-complete":
+                def matches(tags, tag_filters):
+                    for k, v in tag_filters.items():
+                        if k not in tags or tags[k] != v:
+                            return False
+                    return True
+                tags = make_node_tags(labels, status_tag(pod))
+                if matches(tags, tag_filters):
+                    result.append(pod["metadata"]["name"])
+
+        logger.info("Result is {}".format(result))
+
+        return result
+
+    def terminate_node(self, node_id: str) -> None:
+        """Terminates the specified node."""
+        self.terminate_nodes([node_id])
+
+    def terminate_nodes(self, node_ids: List[str]) -> Dict[str, Dict[str, str]]:
+        with self._lock:
+            # Split node_ids into groups according to node type and terminate them individually.
+            # Note that in most cases, node_ids contains a single element and therefore it is
+            # most likely not worth optimizing this code to batch the requests to the API server.
+            groups = {}
+            label_filters = to_label_selector({"ray.io/cluster": self.cluster_name})
+            pods = self._get("/api/v1/namespaces/default/pods?labelSelector=" + requests.utils.quote(label_filters))
+            for pod in pods["items"]:
+                if pod["metadata"]["name"] in node_ids:
+                    groups.setdefault(pod["metadata"]["labels"]["ray.io/group"], []).append(pod["metadata"]["name"])
+
+            url = "/apis/ray.io/v1alpha1/namespaces/default/rayclusters/{}".format(self.cluster_name)
+            raycluster = self._get(url)
+            for group_name, nodes in groups.items():
+                group_index, group_spec = self._get_worker_group(raycluster, group_name)
+                prefix = f"/spec/workerGroupSpecs/{group_index}"
+                payload = [
+                    {"op": "test", "path": prefix + "/replicas", "value": group_spec["replicas"]},
+                    {"op": "replace", "path": prefix + "/replicas", "value": group_spec["replicas"] - len(nodes)},
+                    {"op": "replace", "path": prefix + "/scaleStrategy", "value": {"workersToDelete": nodes}}
+                ]
+                self._patch(url, payload)
+                self._wait_for_pods(group_name, group_spec["replicas"] - len(nodes))
```

Comment:
I am thinking whether we can do some optimization to terminate all and wait all instead of group by group? WDYT?

Reply:
Yes, that is a great idea. For the first iteration I think it makes sense to focus on code clarity (especially if the performance/scalability difference is not big) but I'll try out how the code looks if I make it part of this PR. Otherwise we can do it in a follow up PR, there are some other optimizations I'd like to play with (e.g. I wonder if we can ship less information in the list pods call and avoid having to ship the full pod information).

Reply:
A very simple optimization would be to move the last line into its own for loop.

Reply:
Well, the last line and the next line.

Reply:
Thanks, I included the optimization suggested now, I think it is worth it and doesn't make the code much less comprehensible. For other optimizations, I created an issue here: https://github.com/ray-project/ray/issues/21656